[DOC-1171] LangChain Framework integration (#28845)
yugabyte/yugabyte-db: docs/content/stable/integrations/langchain.md

The example progresses from basic operations (storing and searching documents) to building a complete RAG chain that combines retrieval and generation. Optionally, it can load real YugabyteDB documentation from the web, split it into manageable chunks, and add it to the vector store to create a knowledge base that the LLM can query.

### Prerequisites

- Python 3.9 or later
- Docker
- Create an [OpenAI API key](https://platform.openai.com/api-keys). Export it as an environment variable with the name `OPENAI_API_KEY`.

### Setup

1. Create a virtual environment:

    ```sh
    python3 -m venv .venv
    source .venv/bin/activate
    ```

1. Install all dependencies, including the `langchain-postgres` package:

    ```sh
    pip install --upgrade --quiet langchain langchain-openai langchain-community langchain-postgres tiktoken psycopg-binary langchain-yugabytedb beautifulsoup4
    ```

1. Start YugabyteDB with vector extension support:

    ```sh
    docker run -d --name yugabyte_node01 --hostname yugabyte01 \
    -p 7000:7000 -p 9000:9000 -p 15433:15433 -p 5433:5433 -p 9042:9042 \
    yugabytedb/yugabyte:2.25.2.0-b359 bin/yugabyted start --background=false \

>     --master_flags="allowed_preview_flags_csv=ysql_yb_enable_advisory_locks,ysql_yb_enable_advisory_locks=true" \

Reviewer: This isn't still a preview flag is it? And advisory locks are are true by default now right

5. Removes table tombstone check for colocated tables, increasing read and write performance.

============================================================
Complete example finished successfully!
You now have a working RAG application with YugabyteDB!
```

## Troubleshooting

### Common issues

- Vector extension not working: Make sure you ran all the verification commands.
- Connection issues: Verify YugabyteDB is running and accessible on `localhost:5433`.
- API key issues: Ensure your OpenAI API key is set correctly.
- Driver compatibility: Use `psycopg` instead of `asyncpg` for better vector support.

### Verification steps

Check if everything is working:

```bash
# Check if YugabyteDB is running
docker ps | grep yugabyte

# Check if vector extension is installed
docker exec -it yugabyte_node01 bin/ysqlsh -h yugabyte01 -c "SELECT * FROM pg_extension WHERE extname = 'vector';"

# Test vector functionality
docker exec -it yugabyte_node01 bin/ysqlsh -h yugabyte01 -c "SELECT '[1,2,3]'::vector;"

# Test Python imports
python -c "from langchain_yugabytedb import YBEngine, YugabyteDBVectorStore; print('Imports successful')"
```

## Additional operations

> The main example demonstrates the core RAG workflow. Here are additional operations you can perform with the vector store:

Reviewer: ```suggestion
The main example demonstrates the core RAG workflow. The vector store also supports a number of additional features.
```

table_name = "chat_history"
YugabyteDBChatMessageHistory.create_tables(sync_connection, table_name)

# Generate a unique session ID
session_id = str(uuid.uuid4())

# Initialize the chat history manager
chat_history = YugabyteDBChatMessageHistory(
    table_name, session_id, sync_connection=sync_connection
)
```

### Add messages

```python
# Add messages to the chat history
chat_history.add_messages(
    [
        SystemMessage(content="You are a helpful assistant."),
        HumanMessage(content="What is YugabyteDB?"),
        AIMessage(content="YugabyteDB is a distributed SQL database."),
    ]
)

# Retrieve all messages
print(chat_history.messages)
```

### Use with RAG chain

Combine chat history with your RAG chain for conversational AI:

```python
# Add user question to history
chat_history.add_user_message("What features are in v2025.2.0.0?")

# Get conversation context
conversation_context = chat_history.messages

# Use RAG chain to get answer
response = rag_chain.invoke("What features are in v2025.2.0.0?")

# Add AI response to history
chat_history.add_ai_message(response)
```

## Learn more

- [Develop applications with AI and YugabyteDB](/stable/develop/AI/)

> - [pgvector extension](/stable/additional-features/pg-extensions/extension-pgvector/)

Reviewer: ```suggestion
- [pgvector extension](../../../additional-features/pg-extensions/extension-pgvector/)
```
